File: Joe Zheng/week 2/74.First_Bad_Version.java
/**
 * public class SVNRepo {
 *     public static boolean isBadVersion(int k);
 * }
 * you can use SVNRepo.isBadVersion(k) to judge whether 
 * the kth code version is bad or not.
*/

public class Solution {
    /*
     * @param n: An integer
     * @return: An integer which is the first bad version.
     */
    public int findFirstBadVersion(int n) {
        // write your code here
        
        // if(possibleSmallest == n ) {
        //     return 0;
        // }
        // if (SVNRepo.isBadVersion(possibleSmallest)) {
        //     return possibleSmallest ;
        // } else {
        //     return findFirstBadVersion(possibleSmallest + (n - possibleSmallest) / 2);
        // }
        
        int start = 1;
        int end = n;
        
        while(start + 1 < end) {
            int mid = start + (end - start) / 2;
            
            if (SVNRepo.isBadVersion(mid)) {
                end = mid;
            } else {
                start = mid;
            }
            
        }
        
        if(SVNRepo.isBadVersion(start)) {
            return start;
        }
        
        // if(SVNRepo.isBadVersion(end)) {
        //     return end;
        // }
        
        return end;


    }
}

File: Joe Zheng/javaOutput/production/jiuzhangAlgo/Joe Zheng/week 2/74. First_Bad_Version.py
"""
class SVNRepo:
    @classmethod
    def isBadVersion(cls, id)
        # Run unit tests to check whether verison `id` is a bad version
        # return true if unit tests passed else false.
You can use SVNRepo.isBadVersion(10) to check whether version 10 is a 
bad version.
"""


class Solution:
    firstBadVersion = 0
    """
    @param: n: An integer
    @return: An integer which is the first bad version.
    """

    def findFirstBadVersion(self, n):
        # write your code here
        # assume the biggest version is 5000
        if(SVNRepo.isBadVersion(self.firstBadVersion)):
            return self.firstBadVersion
        _i = 1
        for _ in range(n):
            if(SVNRepo.isBadVersion(_i) == True):
                # Do a binar search from i-1 to i
                self.firstBadVersion = _i / 2
                self.findFirstBadVersion(_i)
            _i *= 2
            # print(i)


if __name__ == '__main__':
    Solution().findFirstBadVersion(5)
